src/databricks/labs/ucx/hive_metastore/mapping.py

```diff
@@ -103,14 +138,70 @@ def skip_table(self, backend: StatementExecutionBackend, schema: str, table: str
             else:
                 logger.error(nf)
         except BadRequest as br:
             logger.error(br)
+
+    def _validate_source_table(self, table: Table) -> Table | None:
+        result = self._backend.fetch(f"SHOW TBLPROPERTIES `{table.database}`.`{table.name}`")
+        for value in result:
+            if value["key"] == "upgraded_to":
+                logger.info(f"{table.key} is set as upgraded to {value['value']}")
+                return None
+            if value["key"] == self.UCX_SKIP_PROPERTY:
+                logger.info(f"{table.key} is marked to be skipped")
+                return None
+        return table
+
+    def _validate_source_database(self, database: str) -> str | None:
+        describe = {}
+        for value in self._backend.fetch(f"DESCRIBE SCHEMA EXTENDED {database}"):
+            describe[value["database_description_item"]] = value["database_description_value"]
+        if self.UCX_SKIP_PROPERTY in TablesCrawler.parse_database_props(describe.get("Properties", "").lower()):
+            logger.info(f"Database {database} is marked to be skipped")
+            return None
+        return database
+
+    def _get_upgraded_tables(self) -> dict[str, TableInfo]:
+        upgraded_tables = {}
+        for catalog in self._ws.catalogs.list():
+            if not catalog.name:
+                continue
+            for schema in self._ws.schemas.list(catalog_name=catalog.name):
+                if not schema.name:
+                    continue
+                for table in self._ws.tables.list(catalog_name=catalog.name, schema_name=schema.name):
+                    if table.properties is not None and "upgraded_from" in table.properties:
+                        upgraded_tables[table.properties["upgraded_from"].lower()] = table
+        return upgraded_tables
+
+    def _get_validated_databases(self):
+        databases = self._backend.fetch("SHOW DATABASES")
+        tasks = []
+        for (database,) in databases:
+            tasks.append(partial(self._validate_source_database, database))
+
+        skip_databases, errors = Threads.gather("Checking databases for skip property", tasks)
+        if len(errors) > 0:
+            logger.error(f"Detected {len(errors)} while scanning Skipped Databases")
+        return [skip_database for skip_database in skip_databases if skip_database]
+
+    def _get_validated_tables(self, tables_crawler: TablesCrawler, table_keys: list[str]):
+        tasks = []
+        for table_to_check in tables_crawler.snapshot():
+            if table_to_check.key not in table_keys:
+                continue
+            tasks.append(partial(self._validate_source_table, table_to_check))
+
+        validated_tables, errors = Threads.gather("Checking all database properties", tasks)
+        if len(errors) > 0:
+            logger.error(f"Detected {len(errors)} while validating tables")
+        return [table for table in validated_tables if table]
```

Comment:
Aren't we already filtering out nulls before this?

Reply:
Done